### PR TITLE
Add support for FS switches

### PR DESCRIPTION
**Pull request by davidc**
This model supports Fiberstore S3900-24F4S and S5800-8TF12S

FS.com seem to have a completely different on every switch they sell. The existing "gcombnps" model does not work even slightly with these two switches.

These two switches are even different from each other, one has "show system" and one does not. However this model is good enough to work with both.

I'm not sure how best to name it so I just called it "fs".

### CODE REVIEW

**Comment by wk** [comment]
This looks good, in order to go in as a PR it will need:
* An attribution line in `CHANGELOG.md`
* A line documenting support in `Supported-OS-Types.md`

As for the model name, it's often helpful to analyze the firmware blob, output from commands such as `show version` or even run the firmware against a tool such as `binwalk` in order to determine the actual underlying OS. It may be the OS is already supported (by another model), and by documenting the correct underlying OS, additional devices may be supported in the future.

**Comment by wk** [comment]
Hi @davidc! Checking if this is still live for you, and if this PR is expected to see follow-up so it can be merged?

File: lib/oxidized/model/fs.rb [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+  # Author: David Croft <davidc.net>
+  #
+  # tested with:
+  #  - Fiberstore S3900-24F4S
+  #  - Fiberstore S5800-8TF12S
+
+  prompt /^\r?([\w.@()-]+?(\(1-16 chars\))?[#>:]\s?)$/
+  comment '! '
+
+  # handle pager
```
Comment (fadenb) [inline lib/oxidized/model/fs.rb]: Nice work. Just got an FS switch and was looking to integrate it into oxidized :)

I believe the pager handling can be simplified a bit by executing `terminal length 0` which seems to disable the pager:
```
ssh *****************.omniit.dev 
XXXXXXXX@192.168.1.11's password: 

      CLI session with the S3900-24T4S is opened.
      To end the CLI session, enter [Exit].

*****************.omniit.dev-1#terminal length ?
  0         No pausing
  <24-200>  Number of lines on screen
*****************.omniit.dev-1#terminal length 0
*****************.omniit.dev-1#show running-config 
Building running configuration. Please wait...
!<Version>0.0.1</Version>
!<stackingDB>000000000000</stackingDB>
!<stackingMac>0Y_YY-YY-YY-YY-YY-YY_YY</stackingMac>
!<stackingMac>00_00-00-00-00-00-00_00</stackingMac>
!<stackingMac>00_00-00-00-00-00-00_00</stackingMac>
!<stackingMac>00_00-00-00-00-00-00_00</stackingMac>
!<stackingMac>00_00-00-00-00-00-00_00</stackingMac>
!<stackingMac>00_00-00-00-00-00-00_00</stackingMac>
!
....... (no pager :) )
```

Comment (davidc) [reply lib/oxidized/model/fs.rb]: Yes, actually I added the pager bit before I later added 'term len 0' at login. But I left the pager swallower in place just in case as I don't entirely trust the quality of the engineering of this software, and some models may not support it.

**Comment by davidc** [comment]
> Hi @davidc! Checking if this is still live for you, and if this PR is expected to see follow-up so it can be merged?

Hi,
yes I am off for a couple of weeks but I am then planning to try and figure out the underlying OS as discussed above and then update this PR.
David

**Comment by davidc** [comment]
I've extracted the firmware for both and they're really quite different.

S5800 has a bunch of files that are copyright "Centec Networks Inc." and there is also a directory /centec_switch. The CLI appears to be implemented in a lot of Centec-copyright Python.

S3900 has little to identify it apart from the compiler for all the executables being "gcc (Broadcom Linux)". And with source files like this:
/home/manager/work/gitproj/tws-169/tws/tws/src/kernel/driver/switch/broadcom/sdk_6.4.1/include/sal/core
/opt/toolchain/iProcLDK/3.4.6/iproc/buildroot/host/usr/arm-broadcom-linux-uclibcgnueabi/sysroot/usr/include
which looks to be an SDK for a Broadcom "iProc" SoC. The CLI is provided by a compiled program /usr/bin/cli. It may be barely-modified stock Broadcom firmware.

So it looks like they have nothing in common with each other besides having enough of an "industry-standard" CLI to handle basic commands like "sh run" and "sh ver".

I can't see any existing models that relate to either. Should I make separate models for each?

**Comment by wk** [comment]
Thanks for looking into this and the excellent detail.

In the case of the FS.com S5800 / Centec, Centec Networks is a whitebox provider with a range of products and their own OS, [CNOS](http://www.centecnetworks.com/en/ProductList.asp?ID=291) - is the CLI you see on this device similar to what's documented in Centec's documentation [here](http://www.centecnetworks.com/en/DownList.asp?ID=9&s0=114&s1=116&s2=154&sr=CLI)? If so, that'd be very interesting to have as a standalone `cnos` model, with a small comment at the top of the model file explaining what this is, and `Supported-OS-Types.md` links to that model for all the CNOS devices (the whitebox Centec boxes), as well as the FS.com box - this way you introduce support for an entire set of vendor devices, plus your own (whitelabeled) FS.com device.

In the case of the S3900, that unfortunately does not look like enough information to trace the device to an OS or a whitebox source, iProc is a low-level SDK that is present on a broad range of products. If shaking the binary blob does not yield additional information, and googling for more peculiar commands from its CLI autocomplete also fails to surface a convincing origin - introducing that as and `fs` model seems like the way to go.

**Comment by wk** [comment]
> with a small comment

For inspiration on what that could look like, take a peek at the `dcnos.rb` model, another Chinese whitelabel vendor, which a user discovered under a brand retailed in Russia by NAG under the name of `SNR`.

**Comment by fadenb** [comment]
I did a few tests with the code from the PR:
* On a `S5800` version `5.3.9`
  * There is apparently no `show system` command. This results in a line containing
  ```! % Incomplete command```
  towards the end of the config.
  * System uptime from the `show version` command is not filtered:
  ```! core01 uptime is 0 days, 0 hours, 36 minutes ```
  resulting in a change every run
* On a `S3900` using version `1.6.9`
  * System uptime from the `show system` command is not filtered:
  ```!  System Up Time         : 0 days, 0 hours, 46 minutes, and 45.2 seconds```
  resulting in a change every run

That aside it works as expected :+1:

**Comment by freddy36** [comment]
FS doesn't develop any of their switches on their own and as you already noticed the corresponding software keeps changing completely.
For the S3900 Series the hardware/software appears to be from Digisol, the manual and software of the S3900-24F4S still contains references to the DG-GS4628T model.
Overall the software is very similar to EdgeCOS from edgecore.
For the S3900 you can use the existing edgecos model, it's working fine except the "show watchdog" command which isn't available on the S3900.

They once told me that the S5800/S5850 is based on IPFusion ZebOS but I'm not sure about it.

If you're going to add support for the different fiberstore switches I definetly would recommend that you add seperate models for each series, e.g. a "FSOSS3900" and a "FSOSS5800".

**Comment by candlerb** [comment]
@freddy36 said:
> For the S3900 you can use the existing edgecos model, it's working fine except the "show watchdog" command which isn't available on the S3900.

I am testing with a new S3900-48T6S-R (firmware 2.2.0E Build 88726), and the edgecos driver, with oxidized 0.28.0.

The main problem is that it doesn't send the "enable" command, which results in the following "config" being retrieved and stored:

```
show running-config
show running-config
^
Unknown command
! show system
! show system
! ^
! Unknown command
! show version
! show version
! ^
! Unknown command
! show watchdog
! show watchdog
! ^
! Unknown command
! show interfaces transceiver
! show interfaces transceiver
! ^
! Unknown command
```

My solution is to patch the driver to be able to send `enable`:

```
--- lib/oxidized/model/edgecos.rb.orig	2021-11-05 22:25:07.334515164 +0000
+++ lib/oxidized/model/edgecos.rb	2021-11-05 23:10:21.733126440 +0000
@@ -43,6 +43,15 @@
   end

   cfg :telnet, :ssh do
+    post_login do
+      if vars(:enable) == true
+        cmd "enable"
+      elsif vars(:enable)
+        cmd "enable", /^[pP]assword:/
+        cmd vars(:enable)
+      end
+    end
+
     post_login 'terminal length 0'
     pre_logout 'exit'
   end
```

and set enable to `true` (the default S3900 config has `aaa authentication enable default none` which accepts the "enable" command without any password)

Now it works partly; the commands "show running-config" and "show version" work fine, but it still doesn't understand all the commands:

```
...
! show system
! Incomplete command
...
! show watchdog
! show watchdog
!      ^
! Too many parameters
! show interfaces transceiver
! show interfaces transceiver
!      ^
! Too many parameters
```

This suggests to me that an "FSOSS3900" device type is still probably worthwhile.

Here's the full "show version" output:

```
#show version
Fiberstore Co., Limited Internetwork Operating System Software
S3900-48T6S-R Series Software, Version 2.2.0E Build 88726, RELEASE SOFTWARE
Copyright (c) 2021 by FS.COM All Rights Reserved
Compiled: 2021-9-16 17:47:8 by SYS, Image text-base: 0x80010000
ROM: System Bootstrap, Version 0.1.8,hardware version:A
Serial num:XXXXXXXX, ID num:XXXXXXXX
System image file is "Switch.bin"
FS S3900-48T6S-R RISC
524288K bytes of memory,16384K bytes of flash
Base ethernet MAC Address: 64:9d:99:XX:XX:XX
PCB version:A
snmp info:
  vend_ID:52642   product_ID:2039   system_ID:1.3.6.1.4.1.52642.1.2039.0
XXXXXXXX uptime is 0:05:54:16, The current time: 2021-11-5 23:17:42
 Reboot history information:
  No. 1: System is rebooted by power-on
  No. 2: System is rebooted by command at 2000-1-2 5:2:49, uptime 1:05:02:28
```

**Comment by candlerb** [comment]
Another problem with S3900 is that the "show version" output includes the current time and uptime, this needs to be filtered out with something like this:

```
  cmd 'show version' do |cfg|
    cfg.gsub! /(\s*uptime is\s*)[0-9:]+/, '\1 <hidden>'
    cfg.gsub! /(\s*current time:\s*)[0-9-]+\s+[0-9:]+/, '\1 <hidden>'
    comment cfg
  end
```

**Comment by freddy36** [comment]
@candlerb

I'm 99% sure that the FS S3900-R Series Switches are manufactured by BDCOM.

The specs and manuals are identical. 
Fiberstore S3900-48T6S-R = BDCOM S2954 (https://www.bdcom.cn/en/cpzxRF/info_50.aspx?itemid=3374).

The BDCOM syntax reminds me on the IOS-XR syntax. You might want to check the iosxr module (https://github.com/ytti/oxidized/blob/master/lib/oxidized/model/iosxr.rb).

If that doesn't work very well you should create a BDCOM module and update the documentation accordingly.

**Comment by candlerb** [comment]
Thank you.

It looks like `iosxr.rb` won't work.  It sends [these commands](https://github.com/ytti/oxidized/blob/master/lib/oxidized/model/iosxr.rb#L35-L47) before `enable`:

```
ldex-r18-sw2>terminal length 0
terminal length 0
^
Unknown command
ldex-r18-sw2>terminal width 0
terminal width 0
^
Unknown command
ldex-r18-sw2>terminal exec prompt no-timestamp
terminal exec prompt no-timestamp
^
Unknown command
ldex-r18-sw2>
```

It [doesn't support](https://github.com/ytti/oxidized/blob/master/lib/oxidized/model/iosxr.rb#L40-L43) sending "enable" without a password (i.e. enable=true)

It sends [these](https://github.com/ytti/oxidized/blob/master/lib/oxidized/model/iosxr.rb#L17-L23) unsupported commands:

```
ldex-r18-sw2#show inventory
show inventory
     ^
Too many parameters
ldex-r18-sw2#show platform
show platform
     ^
Too many parameters
```

BDCOM S2954 does indeed look very similar to S3900-48T6S-R.  The BDCOM has only a single PSU, but has a blanking plate where the second power input would go (rear image available in the "High-res photos" bundle)

I'll look into making another model.

**Comment by candlerb** [comment]
I have made a bdcom driver in PR #2395

However, the S3900-48T6S-R I am testing with is clearly not the same as @davidc's S3900-24F4S in this PR (#1855), since his driver sends `show system` and my switch doesn't accept that (only `show system mtu`)

@davidc: I'd be interested to see the `show version` output from your switch, to compare to mine shown [above](https://github.com/ytti/oxidized/pull/1855#issuecomment-962274623).

**Comment by freddy36** [comment]
This is from a S3900-24T4S:
```
switch-1#show system
 System Up Time         : 112 days, 19 hours, 12 minutes, and 26.14 seconds
 System Name            : <removed>
 System Location        : <removed>
 System Contact         : <removed>
 MAC Address (Unit 1)   : 64-9D-99-16-27-23
 System OID String  : 1.3.6.1.4.1.52642.2.1.45.101
 http Server        : Enabled
 http Server Port   : 80
 https Server       : Enabled
 https Server Port  : 443
 Telnet Server      : Enabled
 Telnet Server Port : 23

System Temperature:
-----------------------------------------------------------------
unit 1     Temperature 1:  50 degrees


System Power status:
-----------------------------------------------------------------
 Unit 1     Main Power Status      : Up
switch-1#show version
Unit 1 -----------------------------
 Serial Number          : CG2002249421N0192
 Hardware Version       : 1.0
 Number of Ports        : 28
 Loader Version         : 1.4
 Operation Code Version : 1.7.0
```

**Comment by davidc** [comment]
> @davidc: I'd be interested to see the `show version` output from your switch, to compare to mine shown [above](https://github.com/ytti/oxidized/pull/1855#issuecomment-962274623).

Yes it's hilariously different

```
switchname-1#sh ver
Unit 1 -----------------------------
 Serial Number          : CG1806nnnnnnlnnn
 Number of Ports        : 28
 Loader Version         : 1.0
 FSOS Software Version  : 1.6.7
 Hardware Version       : 1.0

switchname-1#sh inv
                    ^
% Invalid input detected at '^' marker.
switchname-1#sh system
 FS-SW Up Time          : 128 days, 2 hours, 0 minutes, and 55.81 seconds
 System Name            : switchname
 System Location        : redacted
 System Contact         : redacted
 MAC Address (Unit 1)   : EC-D6-8A-nn-nn-nn
 System OID String  : 1.3.6.1.4.1.52642.2.1.45.103
 http Server        : Enabled
 http Server Port   : 80
 https Server       : Enabled
 https Server Port  : 443
 Telnet Server      : Enabled
 Telnet Server Port : 23

System Temperature:
-----------------------------------------------------------------
unit 1     Temperature 1:  30 degrees


System Power status:
-----------------------------------------------------------------
 Unit 1     Main Power Status      : Up

Copyright (C) Fiberstore Inc.  All rights reserved.

```

Looks more like @freddy36's except even the "System Up Time" is "FS-SW Up Time" instead and the names of the versions are different.

**Comment by candlerb** [comment]
It looks like S3900 and S3900-R are very different models of switch. There's a [comparison document](https://img-en.fs.com/file/competitive_comparison/s3900-r-series-and-s3900-series-switches-competitive-comparison.pdf) which suggests this too.

Here's the sysObjectId from my S3900-R:

```
# snmpbulkwalk ldex-r18-sw2 sysObjectId
SNMPv2-MIB::sysObjectID.0 = OID: SNMPv2-SMI::enterprises.52642.1.2039.0
```

The private enterprise number is the same (52642 = FS.COM INC) but a different subtree (.1 vs .2)

Walking the MIB, I see entries under 52642.2, 52642.3 and 52642.9.  This disagrees with the S3900 MIB file that I [downloaded](https://img-en.fs.com/file/user_manual/fs-s3900-series-switches-snmp-mib-files.zip) from fs.com site, which has:

```
fs              OBJECT IDENTIFIER ::= { enterprises 52642 }
fsnetworks      OBJECT IDENTIFIER ::= { fs 2 }
fsNetworksMgt OBJECT IDENTIFIER ::= { fsnetworks 1 }

fsMIB MODULE-IDENTITY
        LAST-UPDATED "201607150000Z"
        ORGANIZATION "FS.com Inc."
        CONTACT-INFO
                "FS.com Inc.
                 Customer Service"

        DESCRIPTION
                "The MIB module for fs-Series of FS.com Inc."
        REVISION "201707250000Z"
        DESCRIPTION
                "Initial version of this MIB."
        ::= { fsNetworksMgt 45 }

fsMIBObjects       OBJECT IDENTIFIER ::= { fsMIB 1 }
```

In other words, everything should be under 1.3.6.1.4.1.52642.2.1.45.1 and that's definitely not the case for the S3900-R I have here.

**Comment by freddy36** [comment]
@candlerb you might want to ask FS to fix the MIBs or the firmware while they are still selling it. Don't expect and support after that point.

More oddities: according to the FS/BDCOM datasheet it should have 256MB DRAM/32MB flash. According to your output it has 512MB DRAM/16MB flash.

**Comment by candlerb** [comment]
FYI: credit to fs.com, they promptly provided the MIB bundle `S3900-R Switch SNMP MIB Files.rar`:

```
NMS-1705.mib                NMS-DHCP-SERVER-MIB.my  NMS-HAL-TRAP.MIB          NMS-IPAcl.my            NMS-NAT.my               NMS-QOS-PIB-MIB.my      NMS-SNTP.MIB         NMS-WAN-MIB.my
NMS-ACCOUNTING-MIB.my       NMS-EAPS-MIB.my         NMS-IEEE8023-LAG-MIB.my   NMS-IP-ADDRESS-MIB.mib  NMS-NTP-MIB.my           NMS-QOS-QUEUE.my        NMS-SWITCH-MIB.my    NMS-WLAN-MIB.my
NMS-AUTHENTICATION-TRAP.my  NMS-ERPS-MIB.my         NMS-IF-APP-POLICY-MIB.my  NMS-IPSLA-MIB.my        NMS-OPTICAL-PORT-MIB.my  NMS-REMOTE-PING-MIB.my  NMS-SYS.my
NMS-CARD-OPERATION.my       NMS-ETHERRING-MIB.my    NMS-IF-MIB.my             NMS-LLDP-MIB.MIB        NMS-POE-MIB.mib          NMS-ROUTING-MIB.my      NMS-SYS-TRAP.my
NMS-CARD-SYS-MIB.my         NMS-FAN-TRAP.my         NMS-IF-QOS-POLICY.MIB     NMS-LOG-SERVER-MIB.my   NMS-POWER-EXT-MIB.my     NMS-R-QOS-MIB.my        NMS-TC.my
NMS-CFGAPP-MIB.my           NMS-FILTER-TRAP.my      NMS-IF-THRESHOLD-MIB.my   NMS-MacAcl.my           NMS-POWER-MIB.my         NMS-SERIAL.mib          NMS-TRAP.mib
NMS-CHASSIS-MIB.my          NMS-FLASH.my            NMS-INTERFACE-EXT.my      NMS-MEAPS-MIB.my        NMS-PROCESS-MIB.my       NMS-SMI.my              NMS-TS.my
NMS-CONFIG-MGMT.my          NMS-GBSC-MIB.my         NMS-INTERFACES.my         NMS-MEMORY-POOL-MIB.my  NMS-QOS-MIB.my           NMS-SNMP.my             NMS-VLAN-EXT-MIB.my
```

There are a load of bugs in these MIB files, but I've patched them up as well as I can.  And there are still some subtrees which aren't defined in any of these MIB files, e.g. `{ nmslocal 238 }` and `{ nmslocal 239 }`

```
NMS-SMI::nmslocal.238.1.1.1.1 = INTEGER: 0
NMS-SMI::nmslocal.238.1.1.2.1 = STRING: "default"
NMS-SMI::nmslocal.238.1.1.3.1 = IpAddress: 0.0.0.0
NMS-SMI::nmslocal.238.1.1.4.1 = Timeticks: (248691464) 28 days, 18:48:34.64
NMS-SMI::nmslocal.239.1.0 = INTEGER: 0
NMS-SMI::nmslocal.239.2.0 = INTEGER: 0
NMS-SMI::nmslocal.239.3.0 = INTEGER: 0
```

But it's better than no MIB files.

**Review by tomehb — APPROVED** [review]

**Comment by oam7575** [comment]
Hi All.

This PDF from FS.com [s3900-series-switches-fsos-software-release-notes](https://img-en.fs.com/file/user_manual/s3900-series-switches-fsos-software-release-notes.pdf)

Suggests that as of FSOS 1.7.1 ( Oct 2020 ) the CLI for certain modes of switches has been reorganized in an effort to standardize the CLI across the FS switch range.

It may be worth while looking into this before the patch is finalized; just to make sure that FS haven't shifted goal posts.

FS.com [Download page](https://www.fs.com/au/products_support.html?category_type=1&page=1&keyword=s3900&category_software=1)

**Comment by candlerb** [comment]
One problem here is that S3900 and S3900R are *completely* different switches with a *completely* different OS.  The S3900R switches here have a CLI that's substantially different from the one documented for the S3900.  Hence I don't think you can talk about "FSOS" without qualifying it with the actual model or series.

A merge was already accepted for a "bdcom" driver, on the assumption that the S3900R is based on bdcom's series of switches.

**Comment by mortzu** [comment]
Any updates?
